Fix #959 - Add Webjar support

timo-schmid opened the pull request:
Here's my implementation for the Webjar support  (#959) - let me know what you think!

rossabaker commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,104 @@
+
+    // Intercepts the routes that match webjar file names
+    Option(request.pathInfo)
+        .filter(_.startsWith(config.pathPrefix))
```
Seems like we might want some slash-related logic in here.  If the prefix is "foo" and the pathInfo is "foobar", do we want to search for resource "bar"?

`
prefix = "foo"
pathInfo 
`

timo-schmid replied:
I'll have a look at it tomorrow, it's starting to get late here - I'm not too happy with the method anyway, maybe i can split it up to make more sense.

rossabaker commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,104 @@
+    * @param version The version of the webjar
+    * @param asset The asset name inside the webjar
+    */
+  case class WebjarAsset(library: String, version: String, asset: String, pathPrefix: String) {
```
final case classes are good practice.

timo-schmid replied:
TIL :+1:

rossabaker commented on docs/src/main/tut/static.md:
````diff
@@ -44,4 +44,34 @@ val service = HttpService {
 }
 ```
 
+## Webjars
````
Ooh, thank you.  I :heart: pull requests with docs!

timo-schmid commented:
You're right, there's not really a difference. I started renaming from File -> Asset but wasn't completely consistent.

bryce-anderson commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,104 @@
+      * @return The full name in the Webjar
+      */
+    private[staticcontent] def pathInJar: String =
+      s"/META-INF/resources/webjars/$library/$version/$asset"
```
Could we make these strings `val`s? (this and the one above)

timo-schmid replied:
Yeah, vals make a lot more sense here!

bryce-anderson commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,104 @@
+    * @param webjarAsset The WebjarAsset
+    * @return Either the the Asset, if it exist, or Pass
+    */
+  private def serveWebjarAsset(request: Request, webjarAsset: WebjarAsset, config: Config): Option[Task[MaybeResponse]] =
```
I don't think this needs to return an `Option` when its always going to be a `Some`.

timo-schmid replied:
You're right, this is a leftover, thanks for pointing it out!

timo-schmid commented:
@bryce-anderson I made changes according to you suggestions, thanks!

@rossabaker I believe I found what you meant with the slash logic - a path like `prefixwebjar/v1.0/file.js` was actually served. I added a test and fixed it. I also reordered some args in the methods to make the code a bit more readable.

reactormonk commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,142 @@
+      *
+      * @return The path where the asset is reachable on the web
+      */
+    private lazy val webPath: String =
```
Where is this used?

timo-schmid replied:
It could be used in the callbacks to filter files - but it's just a convinience. I can remove it.

reactormonk commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,142 @@
+    * @param cacheStrategy strategy to use for caching purposes. Default to no caching.
+    */
+  final case class Config(filter: WebjarAssetFilter = _ => true,
+                          pathPrefix: String = "",
```
The `pathPrefix` functionality already exists within `mountService`. If you see the need for the functionality for prefixing, I'd prefer `service.contramap(Prefix(...))` so it's available to any service. Or did I miss some reverse routing?

timo-schmid replied:
Yeah, I basically tried to go along with the other 2 services (which both have it) in staticcontent - but I think it could be scratched.

reactormonk commented on server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala:
```diff
@@ -0,0 +1,142 @@
+    */
+  final case class Config(filter: WebjarAssetFilter = _ => true,
+                          pathPrefix: String = "",
+                          executor: ExecutorService = Strategy.DefaultExecutorService,
```
Compiles without the `executor` - are you sure it's necessary?

timo-schmid replied:
I'll remove it!

timo-schmid commented on docs/src/main/tut/static.md:
```diff
@@ -44,4 +44,34 @@ val service = HttpService {
+import org.http4s.server.staticcontent.{WebjarService, webjarService}
+
+// starts the server
+object WebServer extends ServerApp {
```
I'll leave it out, you're right it's not consistent. Any reason to use `Service` instead of `HttpService`, or is it not important which of the 2 to use?

reactormonk replied:
`HttpService` is a more specialized `Service`. Probably better.

timo-schmid commented:
Removing the `pathPrefix` from Config as @reactormonk suggested means that I can get rid of some code, which is cool.

Just for my understanding - what is it used for in the other services in `staticcontent`, if `mountService` provides this feature already?

reactormonk commented:
... I didn't realize the other had `pathPrefix` too. Not sure, might be worth refactoring?

rossabaker commented:
`mountService`, `Router`, and `URITranslation` (which should really be `UriTranslation`) are all means to relocate a service underneath another.  Multiple `mountService` calls is natural in servlet containers, but Blaze assembles them into one service.

I think this should be consistent with the other two.  I guess the question is a breaking change vs. something we don't strictly need.  I could argue this either way.  Does anybody feel strongly?

reactormonk commented:
All of `Router`, `URITranslation` and `mountService` are `Service => ...`, whereas this one is `Service`, so for less-ways-to-do-a-thing, I'd add a deprecation notice with a link to the other ones.
